File: Ruby/Complex.rb
class Complex

   def initialize(real, imag)
      @real, @imag = real, imag
   end

   def Re
      @real
   end

   def Im
      @imag
   end

   # overload assignment operators
   def Re=(real)
      @real = real
   end

   def Im=(imag)
      @imag = imag
   end

   def set(real, imag)
      @real, @imag = real, imag
   end

   # define show
   def show
      print "(", @real, ", ", @imag, ")\n"
   end

   # overload arithmetic operators
   def +(z)
      Complex.new(@real + z.Re, @imag + z.Im)
   end

   def -(z)
      Complex.new(@real - z.Re, @imag - z.Im)
   end

   def *(z)
      Complex.new(@real * z.Re - @imag * z.Im,
                  @real * z.Im + @imag * z.Re)
   end

   # overload comparison operators

end

def testAssignmentOperator
   z = Complex.new(0.0, 0.0)
   z.Re, z.Im = 1.0, 2.0
   z.show
   z.set(3.0, 4.0)
   z.show
   w = z
   w.show
end

def testArithmeticOperators
   z = Complex.new(1.0, 2.0)
   w = Complex.new(3.0, 4.0)
   sum = z + w
   sum.show
   diff = z - w
   diff.show
   prod = z * w
   prod.show
end

testAssignmentOperator
puts
testArithmeticOperators

File: Python/Complex.py
"""
@package Complex
"""

class Complex :

   def __init__(self, real, imag) :
      """
      initialize complex number
      """
      self.real = real
      self.imag = imag

   def Re(self) :
      """
      return real part
      """
      return self.real

   def Im(self) :
      """
      return imaginary part
      """
      return self.imag

   def set(self, real, imag) :
      self.real = real
      self.imag = imag

   def show(self) :
      print "(", self.real, ", ", self.imag, ")"

   def __add__(self, z) :
      return Complex(self.real + z.Re(), self.imag + z.Im())

   def __sub__(self, z) :
      return Complex(self.real - z.Re(), self.imag - z.Im())

   def __mul__(self, z) :
      return Complex(self.real * z.Re() - self.imag * z.Im(), \
                     self.real * z.Im() + self.imag * z.Re())

def testAssignmentOperator() :
   z = Complex(0.0, 0.0)
   z.show()
   z.set(3.0, 4.0)
   z.show()
   w = z
   w.show()

def testArithmeticOperators() :
   z = Complex(1.0, 2.0)
   w = Complex(3.0, 4.0)
   sum = z + w
   sum.show()
   diff = z - w
   diff.show()
   prod = z * w
   prod.show()

if __name__ == '__main__' :
   testAssignmentOperator()
   testArithmeticOperators()

